File: Plus_One.cpp
/*66. Plus One  QuestionEditorial Solution  My Submissions
Total Accepted: 125667
Total Submissions: 350729
Difficulty: Easy
Given a non-negative number represented as an array of digits, plus one to the number.

The digits are stored such that the most significant digit is at the head of the list.

Hide Company Tags Google
Hide Tags Array Math
Hide Similar Problems (M) Multiply Strings (E) Add Binary (M) Plus One Linked List
*/



/*start from the end of vector, if it is 9, change it to 0, 
if it is less than 9, plus one , and return the vector
if run to the end of 0, change vector[0] = 1, push [0] at the end of vector*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int size = digits.size();
        for(int i = size-1; i >= 0; --i){
            if(digits[i] < 9){
                digits[i]++;
                return digits;
            }
            digits[i] = 0;
        }
        
        digits[0] = 1;
        digits.push_back(0);
        return digits;
    }
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int[] plusOne(int[] digits) {
        int size = digits.length;
        //remember from size-1 to 0,
        for(int i = size-1; i >= 0; --i){
            if(digits[i] < 9){
                digits[i]++;
                return digits;
            }
            digits[i] = 0;
        }
        int[] ret = new int[size+1];
        ret[0] = 1;
        return ret;
    }
}

File: Plus_One.swift
// Given a non-empty array of digits representing a non-negative integer, plus one to the integer.

// The digits are stored such that the most significant digit is at the head of the list, and each element in the array contain a single digit.

// You may assume the integer does not contain any leading zero, except the number 0 itself.

// Example 1:

// Input: [1,2,3]
// Output: [1,2,4]
// Explanation: The array represents the integer 123.
// Example 2:

// Input: [4,3,2,1]
// Output: [4,3,2,2]
// Explanation: The array represents the integer 4321.

// Solution 1: 
// directly check in the digits index
// if find a digit < 9, directly add it, otherwise, set this digit to 0, and go through next digit
// 
// Time complexity: O(n)
// Space comlexity: O(1)
class Solution {
    func plusOne(_ digits: [Int]) -> [Int] {
        var plusOne = digits
        
        for i in (0..<digits.count).reversed() {
            if digits[i] < 9 {
                plusOne[i] = digits[i] + 1
                return plusOne
            }
            plusOne[i] = 0
        }
        
        plusOne.insert(1, at: 0)
        return plusOne
    }
}